<!DOCTYPE html>
<html dir="ltr" lang="en">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>DeeZ: reference-based compression by local assembly</title>
	<meta name="keywords" content="deez, compression, sam, bam, samtools, nature methods">
	<meta name="description" content="A tool for compressing SAM/BAM files">
  <link rel="stylesheet" href="combo.css">
  <link href='http://fonts.googleapis.com/css?family=Raleway:400,300,700' rel='stylesheet' type='text/css'>
  <link rel="stylesheet" href="//netdna.bootstrapcdn.com/font-awesome/4.2.0/css/font-awesome.min.css">
  
	
</head>


<body>
  <header>
    <div class="container">
      <h1>DeeZ</h1>
      <div id="menu">
      
        
        <a href="#intro">Intro</a>
      
        
        <a href="#usage">Usage</a>
      
        
        <a href="#info">Support</a>
      
      </div>
    </div>
  </header>


 
  <div id="main_content">     
     
      
      <section id="intro">
        <div class="container ">
          <h2 id="so-what-is-deez">So what is DeeZ?</h2>

<p>DeeZ (a.k.a. <u>Dee</u>NA-<u>Z</u>ip) is a tool for compressing SAM/BAM files, or more formally,
a tool which does reference-based compression by local assembly. DeeZ was publised in <a href="http://www.nature.com/nmeth/journal/v11/n11/full/nmeth.3133.html">Nature Methods in November 2014</a>.</p>

<h2 id="how-do-i-get-deez">How do I get DeeZ?</h2>

<p>Just clone our repository and issue <code>make</code> command:</p>

<pre><code>git clone https://github.com/sfu-compbio/deez.git
cd dz &amp;&amp; make -j
</code></pre>

<p>Alternatively, use:</p>

<pre><code>git clone https://bitbucket.org/compbio/dz.git
git clone git://git.code.sf.net/p/deez-compression/deez \
	deez-compression-deez
</code></pre>

<p>If you don’t have git, you can always fetch pre-packaged DeeZ archives:</p>

<ul>
  <li><a href="https://github.com/sfu-compbio/deez/zipball/master">.zip package</a></li>
  <li><a href="https://github.com/sfu-compbio/deez/tarball/master">.tar.gz package</a> </li>
</ul>

<blockquote>
  <p><strong>Note</strong>: You will need at least g++ 4.4 to compile the sources.</p>
</blockquote>

<h2 id="how-do-i-use-deez">How do I use DeeZ?</h2>

<p>DeeZ is invoked as following:</p>

<ul>
  <li>
    <p><strong>Compression</strong></p>

    <pre><code>deez -r [reference] [input.sam] -o [output]
</code></pre>

    <p>This will compress <code>input.sam</code> to <code>input.sam.dz</code>.</p>
  </li>
  <li>
    <p><strong>Decompression</strong></p>

    <pre><code>deez -r [reference] [input.dz] -o [output] ([region])
</code></pre>

    <p>This will decompress <code>input.dz</code> to <code>input.dz.sam</code>. <code>[region]</code> is optional.</p>
  </li>
  <li>
    <p><strong>Random Access</strong></p>

    <p>You can also specify the region of interest while decompressing (i.e. randomly access the region). 
For example, to extract some reads from chr16 to standard output, you should run:</p>

    <pre><code>deez -r [reference] input.dz  -c chr16:15000000-16000000
</code></pre>

    <p>Don’t forget to <a href="http://www.nature.com/nmeth/journal/v11/n11/ris/nmeth.3133.ris">cite us</a> if you use it in your research :)</p>
  </li>
</ul>


        </div>
      </section>
    
      
      <section id="usage">
        <div class="container ">
          <h2 id="parameter-explanation">Parameter explanation</h2>

<ul>
  <li>
    <p><code>--threads, -t [number]</code></p>

    <p>Set up the number of threads DeeZ may use for compression and decompression.</p>

    <p>Default value: <strong>4</strong></p>
  </li>
  <li>
    <p><code>--reference, -r [file|directory]</code></p>

    <p>Specify the FASTA reference file.</p>

    <blockquote>
      <p><strong>Note</strong>: Chromosome names in the SAM and FASTA files must match. Also, instead of one big FASTA file, DeeZ supports reference lookup in the given directory for chr*.fa files, where chr* is the chromosome ID from the SAM file.</p>
    </blockquote>
  </li>
  <li>
    <p><code>--force, -!</code></p>

    <p>Force overwrite of exiting files.</p>
  </li>
  <li>
    <p><code>--stdout, -c</code></p>

    <p>Compress/decompress to the stdout.</p>
  </li>
  <li>
    <p><code>--output, -o [file]</code></p>

    <p>Compress/decompress to the <code>file</code>.</p>
  </li>
  <li>
    <p><code>--lossy, -l</code></p>

    <p>Set lossy parameter for quality lossy encoding (for more information, please check <a href="http://scalce.sourceforge.net/">SCALCE</a>).</p>
  </li>
  <li>
    <p><code>--quality, -q [mode]</code></p>

    <p>If <code>mode</code> is <strong>1</strong> or <strong>samcomp</strong>, DeeZ will use <a href="https://sourceforge.net/projects/samcomp/">sam_comp</a> quality model to encode the qualities. Quality random access is not supported on those files. </p>
  </li>
  <li>
    <p><code>--withflag, -f [flag]</code></p>

    <p>Decompress only mappings which have <code>flag</code> bits set.</p>
  </li>
  <li>
    <p><code>--withoutflag, -F [flag]</code></p>

    <p>Decompress only mappings which do not have <code>flag</code> bits set.</p>
  </li>
  <li>
    <p><code>--stats, -S</code></p>

    <p>Display mapping statistics (needs DeeZ file as input).</p>
  </li>
  <li>
    <p><code>--sort, -s</code></p>

    <p>Sort the input SAM/BAM file by mapping location.</p>
  </li>
  <li>
    <p><code>--sortmem, -M [size]</code></p>

    <p>Maximum memory used for sorting. </p>

    <p>Default value: <strong>1G</strong></p>
  </li>
</ul>


        </div>
      </section>
    
      
      <section id="info">
        <div class="container ">
          <h2 id="contact--support">Contact &amp; Support</h2>

<p>Feel free to drop any inquiry to <a href="&#109;&#097;&#105;&#108;&#116;&#111;:&#105;&#110;&#117;&#109;&#097;&#110;&#097;&#103;&#064;">inumanag at sfu dot oh canada</a> 
or <a href="&#109;&#097;&#105;&#108;&#116;&#111;:&#102;&#104;&#097;&#099;&#104;&#064;">fhach at sfu dot oh canada</a>.</p>

<h2 id="authors">Authors</h2>

<p>DeeZ has been brought to you by:</p>

<ul>
  <li><a href="http://www.cs.sfu.ca/~fhach/personal/">Dr. Faraz Hach</a></li>
  <li><a href="http://www.sfu.ca/~inumanag">Ibrahim Numanagić</a></li>
  <li><a href="http://www.cs.sfu.ca/~cenk/">Dr. S. Cenk Şahinalp</a></li>
</ul>

<p>from the <a href="http://compbio.cs.sfu.ca">Lab for Computational Biology</a> at <a href="http://www.sfu.ca">Simon Fraser University</a>.</p>

<h2 id="support">Support</h2>

<ul>
  <li>
    <p>NSERC Discovery Grant</p>

    <p><img src="http://www.nserc-crsng.gc.ca/_img/logos/overview_eng.jpg" alt="NSERC" /></p>
  </li>
  <li>
    <p>Vanier Canada Graduate Fellowships	</p>

    <p><img src="http://www.ccnmatthews.com/logos/20090429-vanier_200.jpg" alt="Vanier" /></p>
  </li>
</ul>

<h2 id="licence">Licence</h2>

<p>Copyright (c) 2013, 2014, Simon Fraser University, Indiana University Bloomington. All rights reserved.
Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:</p>

<ul>
  <li>Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.</li>
  <li>Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.</li>
  <li>Neither the name of the Simon Fraser University, Indiana University Bloomington nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.</li>
</ul>

<p>THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS “AS IS” AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.</p>


        </div>
      </section>
    
  </div>
    
  (c) 2013 SFU. All rights reserved.
</body>
</html>

