---
title: "Usage"
---

# Usage

<div id="arguments" markdown="1">

### Parameter explanation

- `--threads, -t [number]`

  Set up the number of threads DeeZ may use for compression and decompression.

  Default value: **4**
  
- `--header, -h`

  Outputs the SAM header.

- `--reference, -r [file|directory]`

  Specify the FASTA reference file.
  
  > **Note**: Chromosome names in the SAM and FASTA files must match. Also, instead of one big FASTA file, DeeZ supports reference lookup in the given directory for chr\*.fa files, where chr\* is the chromosome ID from the SAM file.

- `--force, -!`
  
  Force overwrite of exiting files.

- `--stdout, -c`

  Compress/decompress to the stdout.

- `--output, -o [file]`
  
  Compress/decompress to the `file`.

- `--lossy, -l`

  Set lossy parameter for quality lossy encoding (for more information, please check [SCALCE][1]).

- `--quality, -q [mode]`

  If `mode` is **1** or **samcomp**, DeeZ will use [sam_comp][2] quality model to encode the qualities. Quality random access is not supported on those files. 

- `--withflag, -f [flag]`

  Decompress only mappings which have `flag` bits set.

- `--withoutflag, -F [flag]`

  Decompress only mappings which do not have `flag` bits set.

- `--stats, -S`

  Display mapping statistics (needs DeeZ file as input).

- `--sort, -s`

  Sort the input SAM/BAM file by mapping location.

- `--sortmem, -M [size]`

  Maximum memory used for sorting. 
  
  Default value: **1G**

</div>

[1]: http://scalce.sourceforge.net/
[2]: https://sourceforge.net/projects/samcomp/
